msmarco-passage-ranking/track.py

```diff
@@ -1,5 +1,127 @@
+import csv
 import json
+import logging
 import os
+from collections import defaultdict
+
+import pytrec_eval as pe
+
+Qrels = dict[str, dict[str, int]]
+Results = dict[str, dict[str, float]]
+
+
+def calc_ndcg(qrels: Qrels, results: Results, k_list: list) -> dict:
+    for qid, rels in results.items():
+        for pid in list(rels):
+            if qid == pid:
+                results[qid].pop(pid)
+
+    scores = defaultdict(float)
+
+    metrics = ["ndcg_cut"]
+    pytrec_strings = {f"{metric}.{','.join([str(k) for k in k_list])}" for metric in metrics}
+    evaluator = pe.RelevanceEvaluator(qrels, pytrec_strings)
+    pytrec_scores = evaluator.evaluate(results)
+
+    for query_id in pytrec_scores.keys():
+        for metric in metrics:
+            for k in k_list:
+                scores[f"{metric}@{k}"] += pytrec_scores[query_id][f"{metric}_{k}"]
+
+    queries_count = len(pytrec_scores.keys())
+    if queries_count == 0:
+        return scores
+
+    for metric in metrics:
+        for k in k_list:
+            scores[f"{metric}@{k}"] = float(scores[f"{metric}@{k}"] / queries_count)
+
+    return scores
+
+
+def read_qrels(qrels_input_file):
+    qrels = defaultdict(dict)
+    with open(qrels_input_file, "r") as input_file:
+        tsv_reader = csv.reader(input_file, delimiter="\t")
+        for row in tsv_reader:
+            query_id, doc_id, score = row
+            qrels[query_id][doc_id] = int(score)
+    return qrels
+
+
+def generate_weighted_terms_query(text_expansion_field, query_expansion, boost=1.0):
+    return {
+        "query": {
+            "bool": {
+                "should": [
+                    {"term": {f"{text_expansion_field}": {"value": f"{key}", "boost": value}}} for key, value in query_expansion.items()
+                ],
+                "boost": boost,
+            }
+        }
+    }
+
+
+def generate_bm25_query(text_field, query, boost=1.0):
+    return {"query": {"match": {f"{text_field}": {"query": query, "boost": boost}}}}
+
+
+def generate_combine_bm25_weighted_terms_query(
+    text_field, text_expansion_field, query, query_boost, query_expansion, query_expansion_boost
+):
+    return {
+        "query": {
+            "bool": {
+                "should": [
+                    generate_bm25_query(text_field, query, query_boost)["query"],
+                    generate_weighted_terms_query(text_expansion_field, query_expansion, query_expansion_boost)["query"],
+                ]
+            }
+        }
+    }
+
+
+def generate_pruned_query(field, query_expansion, boost=1.0):
+    return {
+        "query": {
+            "weighted_tokens": {
```

Comment:
I've tried running the modified track and noticed `weighted_tokens` query is not available in 8.11.x. Is this planned for 8.13.0? We need to apply the right versioning as per https://esrally.readthedocs.io/en/stable/track.html#custom-track-repositories. The most recent non-master branch in the track repo is `8.11`. We need to bump up to `8.13` before merging this change.

Reply:
Correct, this is for 8.13.0.